### PR TITLE
feat(storage): takeover appendable object

### PR DESCRIPTION
**Pull request by tritone**
Adds a new method `NewWriterFromAppendableObject` which allows the caller to start writing to an appendable-but-unfinalized object in GCS.

Conformance/emulator tests waiting on full support in the testbench for takeover. Integration tests pass locally, but cannot run in kokoro since zonal buckets are not yet available in the region.

Also fixes a bug where Writer.Attrs is not set appropriately for unfinalized uploads in some cases.

### CODE REVIEW

**Review by madisonhall38 — APPROVED** [review]

File: storage/integration_test.go [hunk 1]
```diff
@@ -3233,7 +3215,100 @@ func TestIntegration_WriterAppend(t *testing.T) {
...
+// Writer test for append takeover of unfinalized object.
+func TestIntegration_WriterAppendTakeover(t *testing.T) {
+	t.Skip("b/402283880")
+	ctx := skipAllButBidi(context.Background(), "ZB test")
+	multiTransportTest(ctx, t, func(t *testing.T, ctx context.Context, _, prefix string, client *Client) {
+		h := testHelper{t}
+		bucketName := prefix + uidSpace.New()
+		bkt := client.Bucket(bucketName)
+
+		h.mustCreateZonalBucket(bkt, testutil.ProjID())
```
Comment (BrennaEpp) [inline storage/integration_test.go:3247]: nit: I'd suggest `mustCreateZonalBucket` also handle the deletion of the bucket using `t.Cleanup()`

File: storage/integration_test.go [hunk 2]
```diff
@@ -3233,7 +3215,100 @@ func TestIntegration_WriterAppend(t *testing.T) {
...
+		defer h.mustDeleteBucket(bkt)
+
+		testCases := []struct {
+			name           string
+			finalize       bool
```
Comment (BrennaEpp) [inline storage/integration_test.go:3252]: Looks like this is unused

File: storage/integration_test.go [hunk 3]
```diff
@@ -3233,7 +3215,100 @@ func TestIntegration_WriterAppend(t *testing.T) {
...
+			},
+			{
+				name:           "middle chunk takeover",
+				finalize:       false,
+				content:        randomBytes9MiB,
+				chunkSize:      4 * MiB,
+				takeoverOffset: 6 * MiB,
+			},
+			{
+				name:           "final chunk takeover",
```
Comment (BrennaEpp) [inline storage/integration_test.go:3279]: Is it possible to do a zero byte write with FinalizeOnClose= true  to finalize an object? If so, maybe it would be good to test?

File: storage/integration_test.go [hunk 4]
```diff
@@ -6637,6 +6712,30 @@ func (h testHelper) mustCreate(b *BucketHandle, projID string, attrs *BucketAttr
 	}
 }
 
+func (h testHelper) mustCreateZonalBucket(b *BucketHandle, projID string) {
+	h.t.Helper()
+
+	// Create a bucket in the same zone as the test VM.
+	zone, err := metadata.ZoneWithContext(context.Background())
+	if err != nil {
+		h.t.Fatalf("could not determine VM zone: %v", err)
```
Comment (BrennaEpp) [inline storage/integration_test.go:6721]: This will cause integration tests to fail locally unless on a VM, correct? Should we add a skip/flag to avoid tests that require this similar to `-short`?

File: storage/integration_test.go (hunk 3 above)
Comment (BrennaEpp) [reply storage/integration_test.go:3279]: Also would be good to test an object with takeoverOffset at zero bytes.

File: storage/integration_test.go [hunk 5]
```diff
@@ -3233,7 +3215,100 @@ func TestIntegration_WriterAppend(t *testing.T) {
...
+			t.Run(tc.name, func(t *testing.T) {
+				// Create non-finalized appendable writer and upload first part of content.
+				obj := bkt.Object(tc.name + uidSpace.New()).Retryer(WithPolicy(RetryAlways))
+				defer h.mustDeleteObject(obj)
+				w := obj.If(Conditions{DoesNotExist: true}).NewWriter(ctx)
+				w.Append = true
+				w.FinalizeOnClose = false
+				w.ChunkSize = tc.chunkSize
+
+				h.mustWrite(w, tc.content[:tc.takeoverOffset])
```
Comment (BrennaEpp) [inline storage/integration_test.go:3296]: What happens if the first writer isn't closed (but is appropriately flushed to GCS through Flush or regular buffer flow), and we try to take over? I assume it will work? But if we try to use this first writer again after taking over, what error do we get?

File: storage/storage.go [hunk 6]
```diff
@@ -1241,6 +1241,48 @@ func (o *ObjectHandle) NewWriter(ctx context.Context) *Writer {
...
 }
 
+// NewWriterFromAppendableObject opens a new Writer to an object which has been
+// partially flushed to GCS, but not finalized. It returns the Writer as well
+// as the current end offset of the object. All bytes written will be appended
+// continuing from the offset.
+//
+// Generation must be set on the ObjectHandle or an error will be returned.
+//
+// Writer attributes such as ChunkSize or ChunkRetryDuration can be set on the
```
Comment (BrennaEpp) [inline storage/storage.go:1251]: What about crc32c?

File: storage/integration_test.go [hunk 7]
```diff
@@ -3233,7 +3215,100 @@ func TestIntegration_WriterAppend(t *testing.T) {
...
+				// Takeover and write remainder of content.
+				gen := h.mustObjectAttrs(obj).Generation
+				w2, off, err := obj.Generation(gen).NewWriterFromAppendableObject(ctx)
+				if err != nil {
+					t.Fatalf("NewWriterFromAppendableObject: %v", err)
+				}
+				if off != tc.takeoverOffset {
+					t.Errorf("takeover offset: got %v, want %v", off, tc.takeoverOffset)
+				}
+				w2.FinalizeOnClose = true
```
Comment (BrennaEpp) [inline storage/integration_test.go:3307]: Should we test setting params on the writer?

Comment (BrennaEpp) [inline storage/integration_test.go:3307]: Should we also test doubly-taken over objects (`w2.FinalizeOnClose = false` here and add a second takeover)? Not sure how much value that would add.

**Review by BrennaEpp — COMMENTED** [review]
Overall LGTM, mostly some non-blocking nits.

File: storage/grpc_writer.go [hunk 8]
```diff
@@ -245,6 +278,7 @@ func (s *gRPCAppendBidiWriteBufferSender) sendOnConnectedStream(buf []byte, offs
 		proto.Merge(req, s.firstMessage)
 	}
 
+	// log.Printf("sending msg: send first msg %v, offset %v, flush %v, finishwrite %v, finalize: %v", sendFirstMessage, offset, flush, finishWrite, finalizeObject)
```
[the remark this answers is not in the document]
Comment (tritone) [reply storage/grpc_writer.go]: Done

File: storage/integration_test.go (hunk 2 above)
Comment (tritone) [reply storage/integration_test.go:3252]: Added a test case using this.

File: storage/integration_test.go (hunk 5 above)
Comment (tritone) [reply storage/integration_test.go:3303]: I think this should cause an error in the first writer on Close; I can test it out.

File: storage/integration_test.go (hunk 7 above)
Comment (tritone) [reply storage/integration_test.go]: Discussed offline, yes we should.

Comment (tritone) [reply storage/integration_test.go]: I don't think it should make a difference, this seems like more of a server functionality test?

File: storage/integration_test.go (hunk 3 above)
Comment (tritone) [reply storage/integration_test.go:3279]: added first byte and last byte cases as well as takeover just to finalize.

File: storage/integration_test.go (hunk 7 above)
Comment (tritone) [reply storage/integration_test.go]: Added some tests with a bit more of this.

File: storage/integration_test.go (hunk 1 above)
Comment (tritone) [reply storage/integration_test.go:3247]: This mirrors how h.mustCreate works currently; we can refactor later if desired.

File: storage/storage.go [hunk 9]
```diff
@@ -1273,16 +1274,49 @@ func (o *ObjectHandle) NewWriterFromAppendableObject(ctx context.Context) (*Writ
...
+}
+
+func (opts *AppendableWriterOpts) apply(w *Writer) {
+	if opts == nil {
+		return
+	}
+	w.ChunkRetryDeadline = opts.ChunkRetryDeadline
+	w.ProgressFunc = opts.ProgressFunc
+	w.ChunkSize = opts.ChunkSize
+	w.FinalizeOnClose = opts.FinalizeOnClose
```
Comment (BrennaEpp) [inline storage/storage.go:1317]: Are the defaults for these set elsewhere? I am concerned a user might set a zero value here on some (but not all) of the fields expecting the default behaviour to occur.

File: storage/integration_test.go (hunk 5 above)
Comment (BrennaEpp) [reply storage/integration_test.go:3331]: Did you test this?

**Comment by BrennaEpp** [comment]
I think we are still missing tests for ProgressFunc and Chunksize?

File: storage/storage.go (hunk 9 above)
Comment (tritone) [reply storage/storage.go:1317]: Default for ChunkRetryDeadline is set in OpenWriter. But, I can add checks.

**Review by BrennaEpp — APPROVED** [review]
Thanks Chris
I think it would be good to actually check that the progress func gets called. This is good enough for now.

File: storage/integration_test.go (hunk 5 above)
Comment (tritone) [reply storage/integration_test.go:3340]: Yes, we get a precondition failed error: 

```close write object "first message takeover-20250415-2531366975931-0004" to bucket "golang-grpc-test-20250415-2531366975931-0003": rpc error: code = FailedPrecondition desc = A different writer has become the exclusive writer of this object.```

**Comment by tritone** [comment]
> Thanks Chris I think it would be good to actually check that the progress func gets called. This is good enough for now.

I added this check as well.

**Review by 2FaceS-bit — APPROVED** [review]
